### PR TITLE
Fix MP Config Optional Lists

### LINKED ISSUE #15927: Can't inject optional list with mpConfig-1.x
**Describe the bug**

Can't inject a type like `Optional<List<String>>` using `@ConfigProperty`

**Steps to Reproduce**

Use `mpConfig-1.3`

Use the following injection code:
```java
    private List<String> connectionProps;
    @Inject
    private void setConnectionProps(@ConfigProperty(name = CONNECTION_PROPERTIES_KEY) Optional<List<String>> connectionProps) {
        this.connectionProps = connectionProps.orElse(Collections.emptyList());
    }
```

See the following error in the log:
```
[15/02/2021, 10:53:51:044 GMT] 00000042 com.ibm.ws.microprofile.config.cdi.ConfigCDIExtension        E CWMCG5003E: The [BackedAnnotatedParameter] Parameter 1 of [BackedAnnotatedMethod] @Inject private com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.KafkaTestClientProvider.setConnectionProps(@ConfigProperty Optional<List<String>>) InjectionPoint dependency was not resolved. Error: java.lang.IllegalArgumentException: CWMCG0018E: The java.util.Optional<java.util.List<java.lang.String>> Generic Type Variable for the java.util.List<java.lang.String> conversion type is not supported.
	at com.ibm.ws.microprofile.config.impl.ConversionManager.convert(ConversionManager.java:129)
	at com.ibm.ws.microprofile.config.impl.ConversionManager.convert(ConversionManager.java:104)
	at com.ibm.ws.microprofile.config.impl.AbstractConfig.convertValue(AbstractConfig.java:195)
	at com.ibm.ws.microprofile.config.impl.AbstractConfig.getValue(AbstractConfig.java:174)
	at com.ibm.ws.microprofile.config.impl.AbstractConfig.getValue(AbstractConfig.java:142)
	at com.ibm.ws.microprofile.config.cdi.ConfigProducer.newValue(ConfigProducer.java:48)
	at com.ibm.ws.microprofile.config.cdi.ConfigCDIExtension.processConversionType(ConfigCDIExtension.java:163)
	at com.ibm.ws.microprofile.config.cdi.ConfigCDIExtension.processParameterizedType(ConfigCDIExtension.java:150)
	at com.ibm.ws.microprofile.config.cdi.ConfigCDIExtension.validateConfigProperty(ConfigCDIExtension.java:99)
	at com.ibm.ws.microprofile.config.cdi.ConfigCDIExtension.processInjectionTarget(ConfigCDIExtension.java:83)
```

**Expected behavior**

An `Optional` should be injected.

**Diagnostic information:**
 - OpenLiberty Version: `integration` branch

**Additional context**
This is particularly problematic as `mpConfig-2.0` has removed the ability to inject an empty list as an empty value is now treated the same as a missing value. This 1.x bug makes it very hard to write code that works on both 1.x and 2.0.

### PR DESCRIPTION
**Pull request by tevans78**
Fix the injection of Optional Lists (and Sets)
fixes #15927

### CODE REVIEW

**Comment by tevans78** [comment]
#build

File: dev/com.ibm.ws.microprofile.config.1.1/test/src/com/ibm/ws/microprofile/config/extended/test/WebSphereConfigTest.java [hunk 1]
```diff
@@ -48,39 +40,12 @@ public void testOptionalString() throws NoSuchFieldException, SecurityException
...
-        Optional<String> optValue = (Optional<String>) value;
-        assertTrue(optValue.isPresent());
-        assertEquals("value1", optValue.get());
+        Object value = config.getValue("key5", String.class, true);
+        assertNull("Value should have been null: " + value, value);
     }
 
-    @SuppressWarnings("unchecked")
     @Test
-    public void testOptionalURL() throws NoSuchFieldException, SecurityException {
```
Comment (Joseph-Cass) [inline dev/com.ibm.ws.microprofile.config.1.1/test/src/com/ibm/ws/microprofile/config/extended/test/WebSphereConfigTest.java:62]: What was this test originally for, and why is it being removed?

**Review by Joseph-Cass — APPROVED** [review]
LGTM. I just have a question about the removed test :)

File: dev/com.ibm.ws.microprofile.config.1.1/test/src/com/ibm/ws/microprofile/config/extended/test/WebSphereConfigTest.java (hunk 1 above)
Comment (tevans78) [reply dev/com.ibm.ws.microprofile.config.1.1/test/src/com/ibm/ws/microprofile/config/extended/test/WebSphereConfigTest.java:62]: This was testing that an Optional<URL> could be created using the internal WebSphereConfig interface ... I changed that interface so it isn't possible any more. If an instance of Optional is needed then the standard getOptionalValue method should be used.

File: dev/com.ibm.ws.microprofile.config.1.1.cdi/src/com/ibm/ws/microprofile/config/cdi/ConfigProducer.java [hunk 2]
```diff
@@ -33,22 +35,43 @@
...
         String propertyName = getPropertyName(injectionPoint, qualifier);
-        Object value = newValue(config, propertyName, defaultValue, type, optional);
+
+        Object value = newValue(config, propertyName, defaultValue, ipType);
+
         return value;
     }
 
-    public static Object newValue(Config config, String propertyName, String defaultValue, Type type, boolean optional) {
+    public static Object newValue(Config config, String propertyName, String defaultValue, Type type) {
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.config.1.1.cdi/src/com/ibm/ws/microprofile/config/cdi/ConfigProducer.java:50]: The methods in this class could really have done with some comments to explain what they're expecting and which parts of the config lookup process they handle.

Looking at this PR, it looks like there are several similar-looking methods which each handle different parts of the process and call into each other.

**Pull request merged**